File: src/main/java/qmech/lib/objects/BlockBase.java
package qmech.lib.objects;

import cpw.mods.fml.common.registry.GameRegistry;
import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.item.ItemStack;
import net.minecraft.util.MathHelper;
import net.minecraft.world.World;
import qmech.lib.util.LoggingHelper;
import qmech.mod.ModConfig;
import qmech.mod.Reference;

public class BlockBase extends Block {

    protected static final LoggingHelper logger = LoggingHelper.getInstance();

    public BlockBase(Material material, String intName) {
        super(material);
        this.setBlockTextureName(Reference.MOD_ID + ":" + intName);
        this.setInternalName(intName);
    }

    public static BlockBase config
            (BlockBase block, CreativeTabs ctab, float hardness, float blastResistance, String toolType, int toolLevel) {
        BlockInfo info = ModConfig.getBlockInfo(block, hardness, blastResistance, toolType, toolLevel);
        logger.debug(String.format("Configuring block (%s) with : \n" +
                        ">>> CreativeTab : %s \n" +
                        ">>> Hardness : %s, BlastResistance : %s \n" +
                        ">>> Tool : %s, %s",
                block.getUnlocalizedName(),
                ctab, info.hardness, info.blastResistance, info.toolType, info.toolLevel
        ));
        block.setCreativeTabs(ctab);
        block.setStrength(info.hardness, info.blastResistance);
        block.setHarvestLevel(info.toolType, info.toolLevel);
        block.registerBlock();
        return block;
    }

    public static BlockBase config
            (BlockBase block, CreativeTabBase ctab, BlockInfo info) {
        return config(block, ctab, info.hardness, info.blastResistance, info.toolType, info.toolLevel);
    }

    protected BlockBase setStrength(float hardness, float blastResistance) {
        this.setHardness(hardness);
        this.setResistance(blastResistance);
        return this;
    }

    protected void setCreativeTabs(CreativeTabs creativeTab) {
        this.setCreativeTab(creativeTab);
    }

    void setInternalName(String name) {
        this.setBlockName(name);
    }

    protected void registerBlock() {
        GameRegistry.registerBlock(this, this.getUnlocalizedName().substring(5));
    }

    public void setHarvestLevel(String toolType, int toolLevel) {
        super.setHarvestLevel(toolType, toolLevel);
    }

    public void onBlockPlacedBy(World world, int x, int y, int z, EntityLivingBase player, ItemStack stack) {
        int dir = MathHelper.floor_double((double) ((player.rotationYaw * 4F) / 360F) + 0.5D) & 3;
        world.setBlockMetadataWithNotify(x, y, z, dir, 0);
    }

    public static class BlockInfo {
        public float hardness = 1.0F;
        public float blastResistance = 1.0F;
        public String toolType = "pickaxe";
        public int toolLevel = 1;

        public BlockInfo() {
        }

        public BlockInfo(float hardness, float blastResistance, String toolType, int toolLevel) {
            this.hardness = hardness;
            this.blastResistance = blastResistance;
            this.toolType = toolType;
            this.toolLevel = toolLevel;
        }

        public static BlockInfo ironBlock() {
            return new BlockInfo(5.0F, 10.0F, "pickaxe", 1);
        }
    }
}

File: src/main/scala/qmech/lib/objects/block/BlockBase.scala
package qmech.lib.objects.block

import net.minecraft.block.material.Material
import net.minecraft.block.Block
import net.minecraft.creativetab.CreativeTabs
import cpw.mods.fml.common.registry.GameRegistry
import net.minecraft.world.World
import net.minecraft.entity.EntityLivingBase
import net.minecraft.item.ItemStack
import net.minecraft.util.MathHelper
import qmech.mod.init.{Reference, ModConfigs}
import qmech.lib.objects.info.BlockInfo
import qmech.lib.objects.CreativeTabBase

/**
 * Created by anshuman on 26-05-2014.
 */
class BlockBase(blockMaterial: Material, val intName: String) extends Block(blockMaterial) {
  this.setBlockTextureName(Reference.MOD_ID + ":" + intName)
  this.setInternalName(intName)

  def this(intName: String, blockMat: Material, ctab: CreativeTabs, hardness: Float, blastResistance: Float, toolType: String, toolLevel: Int) {
    this(blockMat, intName)
    val info: BlockInfo = ModConfigs.getBlockInfo(this, hardness, blastResistance, toolType, toolLevel)
    setCreativeTabs(ctab)
    setStrength(info.hardness, info.blastResistance).setHarvestLevel(info.toolType, info.toolLevel)
    registerBlock()
  }

  def this(intName: String, blockMat: Material, ctab: CreativeTabBase, nfo: BlockInfo) {
    this(blockMat, intName)
    val info: BlockInfo = ModConfigs.getBlockInfo(this, nfo.hardness, nfo.blastResistance, nfo.toolType, nfo.toolLevel)
    setCreativeTabs(ctab)
    setStrength(info.hardness, info.blastResistance).setHarvestLevel(info.toolType, info.toolLevel)
    registerBlock()
  }

  def setStrength(hardness: Float, blastResistance: Float): BlockBase = {
    this.setHardness(hardness)
    this.setResistance(blastResistance)
    this
  }

  def setCreativeTabs(creativeTab: CreativeTabs): BlockBase = {
    this.setCreativeTab(creativeTab)
    this
  }

  def setInternalName(name: String): BlockBase = {
    this.setBlockName(name)
    this
  }

  def registerBlock() {
    GameRegistry.registerBlock(this, this.getUnlocalizedName.substring(5))
  }

  override def setHarvestLevel(toolType: String, toolLevel: Int) {
    super.setHarvestLevel(toolType, toolLevel)
  }

  override def onBlockPlacedBy(world: World, x: Int, y: Int, z: Int, player: EntityLivingBase, stack: ItemStack) {
    val dir: Int = MathHelper.floor_double(((player.rotationYaw * 4F) / 360F).asInstanceOf[Double] + 0.5D) & 3
    world.setBlockMetadataWithNotify(x, y, z, dir, 0)
  }

  def newItemStack(size: Int = 1) = new ItemStack(this, size)
}
